File: t/PACSFramework/CS/MPPSWCFDemo/UI/Pages/PatientPage.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using Leadtools.Wizard.Pages;
using Leadtools.Wizard;
using MPPSWCFDemo.Broker;
using Leadtools.Demos;
using System.ServiceModel;
using Leadtools.Dicom;

namespace MPPSWCFDemo.UI.Pages
{
    public partial class PatientPage : InternalTemplatePage
    {       
        public PatientPage()
        {
            InitializeComponent();                     
        }        

        public override void OnSetActive(object sender, WizardCancelEventArgs e)
        {
            base.OnSetActive(sender, e);
            UpdateUI(false);       
        }

        protected override WizardSheet.WizardButtons GetWizardButtons()
        {
            WizardSheet.WizardButtons buttons =  base.GetWizardButtons();

            buttons |= WizardSheet.WizardButtons.Option1 & ~WizardSheet.WizardButtons.Next | WizardSheet.WizardButtons.Finish;
            return buttons;
        }

        public override void OnOptionButton1(object sender, EventArgs e)
        {
            textBoxId.Text = string.Empty;
            textBoxIssuerOfPatientId.Text = string.Empty;           
            dateTimePickerBirth.Value = DateTime.Now;
            dateTimePickerBirth.Checked = false;
            comboBoxSex.SelectedIndex = -1;
            textBoxName.Text = string.Empty;                       
        }

        public override void OnWizardFinish(object sender, CancelEventArgs e)
        {
            ProgressDialog dlgProgresss = new ProgressDialog();
            BrokerServiceClient client = GetWizard().Tag as BrokerServiceClient;

            UpdatePatient();
            dlgProgresss.Title = "Update patient information";
            dlgProgresss.Description = "Updating...";
            dlgProgresss.Action = () =>
            {
                client.UpdateMPPS(MainForm.Mpps.MPPSSOPInstanceUID, MainForm.Mpps);
            };

            if (dlgProgresss.ShowDialog(this) == DialogResult.Cancel)
            {
                if (dlgProgresss.Exception != null)
                {
                    e.Cancel = true;
                    Messager.ShowError(this, dlgProgresss.Exception);
                }
            }

            base.OnWizardFinish(sender, e);
        }                     

        private void UpdateUI(bool clear)
        {
           if (!clear)
           {
              if (MainForm.Mpps.Patient == null)
                 MainForm.Mpps.Patient = new PatientInfoforUnscheduledPPS();

              textBoxId.Text = MainForm.Mpps.Patient.PatientID;
              textBoxIssuerOfPatientId.Text = MainForm.Mpps.Patient.IssuerofPatientID;
              textBoxName.Text = MainForm.Mpps.Patient.PatientName;

              if (MainForm.Mpps.Patient.PatientBirthDate.HasValue)
              {
                 dateTimePickerBirth.Value = MainForm.Mpps.Patient.PatientBirthDate.Value.Date1.ToDateTime();
              }

              if (!string.IsNullOrEmpty(MainForm.Mpps.Patient.PatientSex))
              {
                 int index = comboBoxSex.FindStringExact(MainForm.Mpps.Patient.PatientSex);

                 comboBoxSex.SelectedIndex = index;
              }
              else
                 comboBoxSex.Text = "";
           }
           else
           {
              textBoxId.Text = string.Empty;
              textBoxIssuerOfPatientId.Text = string.Empty;
              textBoxName.Text = string.Empty;
              dateTimePickerBirth.Value = DateTime.Now;
              comboBoxSex.SelectedIndex = -1;
           }
        }

        private void UpdatePatient()
        {
            MainForm.Mpps.Patient.PatientID = textBoxId.Text;
            MainForm.Mpps.Patient.IssuerofPatientID = textBoxIssuerOfPatientId.Text;
            MainForm.Mpps.Patient.PatientName = textBoxName.Text;
            MainForm.Mpps.Patient.PatientSex = comboBoxSex.Text;

            if (dateTimePickerBirth.Checked)
                MainForm.Mpps.Patient.PatientBirthDate = new DicomDateRangeValue() { Date1 = new DicomDateValue(dateTimePickerBirth.Value) };
        }

        public override void OnReset()
        {
           UpdateUI(true);
        }                         
    }
}

File: t/PACSFramework/VB/MPPSWCFDemo/UI/Pages/PatientPage.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports Leadtools.Wizard.Pages
Imports Leadtools.Wizard
Imports MPPSWCFDemo.Broker
Imports Leadtools.Demos
Imports System.ServiceModel
Imports Leadtools.Dicom

Namespace MPPSWCFDemo.UI.Pages
	Partial Public Class PatientPage
		Inherits InternalTemplatePage
		Public Sub New()
			InitializeComponent()
		End Sub

		Public Overrides Sub OnSetActive(ByVal sender As Object, ByVal e As WizardCancelEventArgs)
			MyBase.OnSetActive(sender, e)
			UpdateUI(False)
		End Sub

		Protected Overrides Function GetWizardButtons() As WizardSheet.WizardButtons
			Dim buttons As WizardSheet.WizardButtons = MyBase.GetWizardButtons()

			buttons = buttons Or WizardSheet.WizardButtons.Option1 And (Not WizardSheet.WizardButtons.Next) Or WizardSheet.WizardButtons.Finish
			Return buttons
		End Function

		Public Overrides Sub OnOptionButton1(ByVal sender As Object, ByVal e As EventArgs)
			textBoxId.Text = String.Empty
			textBoxIssuerOfPatientId.Text = String.Empty
			dateTimePickerBirth.Value = DateTime.Now
			dateTimePickerBirth.Checked = False
			comboBoxSex.SelectedIndex = -1
			textBoxName.Text = String.Empty
		End Sub

	   Public _clientPatientPageOnWizardFinish As BrokerServiceClient
		Public Overrides Sub OnWizardFinish(ByVal sender As Object, ByVal e As CancelEventArgs)
			Dim dlgProgresss As New ProgressDialog()
			_clientPatientPageOnWizardFinish = TryCast(GetWizard().Tag, BrokerServiceClient)

			UpdatePatient()
			dlgProgresss.Title = "Update patient information"
			dlgProgresss.Description = "Updating..."
			dlgProgresss.Action = AddressOf AnonymousMethodPatientPageOnWizardFinish

			If dlgProgresss.ShowDialog(Me) = DialogResult.Cancel Then
				If dlgProgresss.Exception IsNot Nothing Then
					e.Cancel = True
					Messager.ShowError(Me, dlgProgresss.Exception)
				End If
			End If

			MyBase.OnWizardFinish(sender, e)
		End Sub
		Private Sub AnonymousMethodPatientPageOnWizardFinish()
			_clientPatientPageOnWizardFinish.UpdateMPPS(MainForm.Mpps.MPPSSOPInstanceUID, MainForm.Mpps)
		End Sub

		Private Sub UpdateUI(ByVal clear As Boolean)
		   If (Not clear) Then
			  If MainForm.Mpps.Patient Is Nothing Then
				 MainForm.Mpps.Patient = New PatientInfoforUnscheduledPPS()
			  End If

			  textBoxId.Text = MainForm.Mpps.Patient.PatientID
			  textBoxIssuerOfPatientId.Text = MainForm.Mpps.Patient.IssuerofPatientID
			  textBoxName.Text = MainForm.Mpps.Patient.PatientName

			  If MainForm.Mpps.Patient.PatientBirthDate.HasValue Then
				 dateTimePickerBirth.Value = MainForm.Mpps.Patient.PatientBirthDate.Value.Date1.ToDateTime()
			  End If

			  If (Not String.IsNullOrEmpty(MainForm.Mpps.Patient.PatientSex)) Then
				 Dim index As Integer = comboBoxSex.FindStringExact(MainForm.Mpps.Patient.PatientSex)

				 comboBoxSex.SelectedIndex = index
			  Else
				 comboBoxSex.Text = ""
			  End If
		   Else
			  textBoxId.Text = String.Empty
			  textBoxIssuerOfPatientId.Text = String.Empty
			  textBoxName.Text = String.Empty
			  dateTimePickerBirth.Value = DateTime.Now
			  comboBoxSex.SelectedIndex = -1
		   End If
		End Sub

		Private Sub UpdatePatient()
			MainForm.Mpps.Patient.PatientID = textBoxId.Text
			MainForm.Mpps.Patient.IssuerofPatientID = textBoxIssuerOfPatientId.Text
			MainForm.Mpps.Patient.PatientName = textBoxName.Text
			MainForm.Mpps.Patient.PatientSex = comboBoxSex.Text

			If dateTimePickerBirth.Checked Then
				MainForm.Mpps.Patient.PatientBirthDate = New DicomDateRangeValue() With {.Date1 = New DicomDateValue(dateTimePickerBirth.Value)}
			End If
		End Sub

		Public Overrides Sub OnReset()
		   UpdateUI(True)
		End Sub
	End Class
End Namespace
